#### Pandas -

- Works with tabular data


- Data Structures in Pandas

  1. Series-  It has a Index and data
  2. Data Frames- Similar to excel / sql table.

In [5]:
# Importing pandas


import pandas as pd

#### Series

####  Syntax- pandas.Series(data=None, index=None, dtype=None, name=None, copy=None, fastpath=<no_default>)


In [6]:
index_list= [0,1,2,3,4,5,6]

data_list= [100,200,300,400,500,600,700]



In [11]:
series_1= pd.Series(data= data_list, index= index_list)

In [19]:
series_1

0    100
1    200
2    300
3    400
4    500
5    600
6    700
dtype: int64

In [24]:

import pandas as pd


month_2021 = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 
              'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

revenue_2021 = [2323, 2445, 234, 3455, 5545, 45456, 
                657, 4566, 54435, 5455, 3555, 6787]

In [27]:
revenue_series = pd.Series(data=revenue_2021, index=month_2021)

In [28]:
revenue_series

Jan     2323
Feb     2445
Mar      234
Apr     3455
May     5545
Jun    45456
Jul      657
Aug     4566
Sep    54435
Oct     5455
Nov     3555
Dec     6787
dtype: int64

#### Dictioneries

#### With dictionaries the keys are read as indexes and the values ar read as data

In [30]:
dict_1={'key1':'value1','key2':'value2','key3':'value3'}

In [31]:
pd.Series(data=dict_1,index=dict_1)

key1    value1
key2    value2
key3    value3
dtype: object

#### Data Frames

#### Syntax-  pandas.DataFrame(data=None, index=None, columns=None, dtype=None, copy=None)

In [32]:
row_labels = [0,1,2,3,4]
column_labels = ['A','B','C']
data = [[1,2,5],(4,2,4),{6,7,9},[3,2,4],[3,3,2]]

In [33]:
# Creating a DataFame using the variables in the previous cell
df = pd.DataFrame(index=row_labels, data=data, columns=column_labels)

In [34]:
df

A  B  C
0  1  2  5
1  4  2  4
2  9  6  7
3  3  2  4
4  3  3  2

#### set_index()

#### The set_index() function can set an index using existing columns

In [35]:
# sets column 'C' as the index
df.set_index('C')

A  B
C      
5  1  2
4  4  2
7  9  6
4  3  2
2  3  3

####  reset_index()

The reset_index function resets the index

In [36]:
df.reset_index(drop=True,inplace=True)

In [38]:
df

A  B  C
0  1  2  5
1  4  2  4
2  9  6  7
3  3  2  4
4  3  3  2

## Project-   Employees Dataset

#### Dataset overview:
There are three csv files containing tables stored in the same directory as this Notebook, they are all related to each other:

• employees.csv: contains information about employees in a company. It contains their unique employee number (emp_no), their department number (dept_no), their hire date (hire_date) and their leaving date (leaving_date). The leaving date is blank if the employee is still employed by the company

• departments.csv: This contains information about the departments in a company. It contains the deparment number (dept_no), the department name (dept_name) and location.

• salaries.csv: This file contains the salaries of the employees. It contains a unique employee department key (emp_dept_key) and the salary. The emp_dept_key is in the format 'emp_id-dept_id'

####  Import pandas, numpy and datetime

In [39]:
import pandas as pd
import numpy as np
import datetime as dt

#### Load the files:
• "employees.csv" should be assigned to the variable emp

• "departments.csv" should be assigned to the variable dept

• "salaries.csv" should be assigned to the variable sal

In [40]:
emp = pd.read_csv('employees.csv')

In [41]:
dept = pd.read_csv('departments.csv')

In [42]:
sal = pd.read_csv('salaries.csv')

##### Head and Tail   

Head- Return the first 5 rows
Tail- Return the last 5 rows

Check the head of all three DataFrames

In [43]:
emp.head()

emp_no dept_no   hire_date leaving_date
0   10001      D5  26/06/2006          NaN
1   10002      D6  21/11/2005          NaN
2   10003      D4  28/08/2006          NaN
3   10004      D4  01/12/2006          NaN
4   10005      D3  12/09/2009          NaN

In [44]:
dept.head()

dept_no                dept_name  location
0      D1   Accounting and Finance   Chicago
1      D2          Human Resources  New York
2      D3  Supply Chain Operations   Chicago
3      D4                Marketing  New York
4      D5               Technology   Chicago

In [45]:
sal.head()

emp_dept_key  salary
0     10001-D5   30546
1     10002-D6   36536
2     10003-D4   38323
3     10004-D4   31851
4     10005-D3   53435

##### Check the data types of all three DataFrames

In [46]:
emp.dtypes

emp_no           int64
dept_no         object
hire_date       object
leaving_date    object
dtype: object

In [48]:
dept.dtypes

dept_no      object
dept_name    object
location     object
dtype: object

In [49]:
sal.dtypes

emp_dept_key    object
salary           int64
dtype: object

#### Change the data types accordingly.
• emp_no, dept_no, dept_name, location, emp_dept_key should all be string data types

• hire_date and leaving_date should be datetime64

• salary should be int64

In [51]:
emp[['emp_no', 'dept_no']]= emp[['emp_no', 'dept_no']].astype('string')

In [53]:
emp['hire_date'] = pd.to_datetime(emp['hire_date'], format = '%d/%m/%Y')

In [56]:
emp['leaving_date']= pd.to_datetime(emp['leaving_date'], format= '%d/%m/%Y')

In [57]:
dept[['dept_no', 'dept_name', 'location']] = dept[['dept_no', 'dept_name', 'location']].astype('string')

In [63]:
sal['emp_dept_key']= sal['emp_dept_key'].astype('string')

#### How many employees are currently working at the company.

The employees still employed, but do not have a leaving date value. 
You can use the isnull() method to identify nulls or NaN values.

In [64]:
# isnull() returns True when the 'leaving date' column contains a null value
emp['leaving_date'].isnull()

0       True
1       True
2       True
3       True
4       True
       ...  
995     True
996     True
997    False
998    False
999     True
Name: leaving_date, Length: 1000, dtype: bool

In [65]:
# We can use the above to perform boolean indexing and filter the records and then count the filtered records
emp[emp['leaving_date'].isnull() == True].count()

emp_no          741
dept_no         741
hire_date       741
leaving_date      0
dtype: int64

So there are 741 employees which are currently working

#### How many currently employed people are there per department.

In [66]:
current_emp = emp[emp['leaving_date'].isnull() == True]

In [68]:
current_emp.groupby(by='dept_no').count()['emp_no']

dept_no
D1     84
D2     36
D3     94
D4    171
D5    188
D6    168
Name: emp_no, dtype: int64

It gives the total emp per dept.

#### Perform a left join on the emp and dept DataFrames (with emp as the left DF). Assign the result of this to the variable emp_dept

In [69]:
emp_dept = pd.merge(left=emp,right=dept,left_on='dept_no', right_on = 'dept_no', how='left')

In [70]:
emp_dept.head()

emp_no dept_no  hire_date leaving_date                dept_name  location
0  10001      D5 2006-06-26          NaT               Technology   Chicago
1  10002      D6 2005-11-21          NaT                    Sales   Chicago
2  10003      D4 2006-08-28          NaT                Marketing  New York
3  10004      D4 2006-12-01          NaT                Marketing  New York
4  10005      D3 2009-09-12          NaT  Supply Chain Operations   Chicago

#### Perform a left join on the newly created "emp_dept" DataFrame and the "sal" DataFrame.¶
Assign this resulting DataFrame to the variable ""

In [71]:
sal.head()

emp_dept_key  salary
0     10001-D5   30546
1     10002-D6   36536
2     10003-D4   38323
3     10004-D4   31851
4     10005-D3   53435

In [72]:
# We can use lambda and apply to slice the emp_dept_key column values
sal['emp_no']=sal['emp_dept_key'].apply(lambda x :x[:5])   ## This returns the first 5 characters from the column emp_dept_key 
                                                           ##( so it will match the col emp_no from emp_dept table)

In [75]:
emp_dept_sal= pd.merge(left= emp_dept, right= sal, left_on= 'emp_no', right_on = 'emp_no', how='left')


In [76]:
emp_dept_sal.head()

emp_no dept_no  hire_date leaving_date                dept_name  location  \
0  10001      D5 2006-06-26          NaT               Technology   Chicago   
1  10002      D6 2005-11-21          NaT                    Sales   Chicago   
2  10003      D4 2006-08-28          NaT                Marketing  New York   
3  10004      D4 2006-12-01          NaT                Marketing  New York   
4  10005      D3 2009-09-12          NaT  Supply Chain Operations   Chicago   

  emp_dept_key  salary  
0     10001-D5   30546  
1     10002-D6   36536  
2     10003-D4   38323  
3     10004-D4   31851  
4     10005-D3   53435

#### Drop the column "emp_dept_key"

In [77]:
emp_dept_sal.drop(columns='emp_dept_key', inplace=True)

In [80]:
emp_dept_sal.head()

emp_no dept_no  hire_date leaving_date                dept_name  location  \
0  10001      D5 2006-06-26          NaT               Technology   Chicago   
1  10002      D6 2005-11-21          NaT                    Sales   Chicago   
2  10003      D4 2006-08-28          NaT                Marketing  New York   
3  10004      D4 2006-12-01          NaT                Marketing  New York   
4  10005      D3 2009-09-12          NaT  Supply Chain Operations   Chicago   

   salary  
0   30546  
1   36536  
2   38323  
3   31851  
4   53435

#### What is the average salary per department?

In [81]:
emp_dept_sal[['dept_name', 'salary']].groupby(by= 'dept_name').mean()

salary
dept_name                            
Accounting and Finance   52148.887324
Human Resources          44849.602740
Marketing                41540.041985
Sales                    59192.053640
Supply Chain Operations  45183.788462
Technology               58975.979592

In [83]:
# via pivot_table
emp_dept_sal.pivot_table(index= 'dept_name', values= 'salary', aggfunc='mean')

salary
dept_name                            
Accounting and Finance   52148.887324
Human Resources          44849.602740
Marketing                41540.041985
Sales                    59192.053640
Supply Chain Operations  45183.788462
Technology               58975.979592

#### What is the average salary by location?

In [84]:
emp_dept_sal[['location', 'salary']].groupby(by= 'location').mean()

salary
location              
Chicago   55383.208675
New York  42261.229851

#### How many people were hired each year in each of the last 10 years?

In [85]:
emp_dept_sal['hire_year']= emp_dept_sal['hire_date'].dt.strftime('%Y')

In [86]:
emp_dept_sal

emp_no dept_no  hire_date leaving_date                dept_name  \
0     10001      D5 2006-06-26          NaT               Technology   
1     10002      D6 2005-11-21          NaT                    Sales   
2     10003      D4 2006-08-28          NaT                Marketing   
3     10004      D4 2006-12-01          NaT                Marketing   
4     10005      D3 2009-09-12          NaT  Supply Chain Operations   
...     ...     ...        ...          ...                      ...   
1183  10905      D4 2005-02-28   2006-03-07                Marketing   
1184  10906      D2 2014-01-20   2021-04-25          Human Resources   
1185  10906      D2 2014-01-20   2021-04-25          Human Resources   
1186  10906      D6 2014-01-20          NaT                    Sales   
1187  10906      D6 2014-01-20          NaT                    Sales   

      location  salary hire_year  
0      Chicago   30546      2006  
1      Chicago   36536      2005  
2     New York   38323      2006  
3     New York   31851      2006  
4      Chicago   53435      2009  
...        ...     ...       ...  
1183  New York   32735      2005  
1184  New York   29095      2014  
1185  New York   97330      2014  
1186   Chicago   29095      2014  
1187   Chicago   97330      2014  

[1188 rows x 8 columns]

In [87]:
emp_dept_sal[['hire_year', 'emp_no']].groupby(by= 'hire_year').count().tail(10)

emp_no
hire_year        
2010          146
2011           99
2012           81
2013           61
2014           69
2015           39
2016           40
2017           14
2018            8
2019            7